keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1197 @@
+}
+```
+
+"ResourceClaimParameters" and "resource.k8s.io" are valid here. A class may
+also allow both this native type and one or more vendor CRDs as parameters.
+
+### NodeResourceSlice
+
+For each node, one or more NodeResourceSlice objects get created. The kubelet
+publishes them with the node as the owner, so they get deleted when a node goes
+down and then gets removed.
+
+```
+// NodeResourceSlice provides information about available
+// resources on individual nodes.
+type NodeResourceSlice struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// NodeName identifies the node where the capacity is available.
+	// A field selector can be used to list only NodeResources
+	// objects with a certain node name.
+	NodeName string
+
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only NodeResources
+	// objects with a certain driver name.
+	DriverName string
+
+	// Resources describes resources managed by the driver on
+	// the node. It is possible to create more than one NodeResources
+	// object for the same node and driver if this array becomes to large
+	// for a single object.
+	Resources []NodeResource
+}
+
+// NodeResource describes one resource.
+type NodeResource struct {
+	// ID is chosen by the driver to distinguish between different resource
+	// instances. It only needs to be unique for the driver and node.  In
+	// other words, the tuple of NodeName, DriverName, InstanceID needs to
+	// be unique in the cluster.
+	ID string
+
+    // Attributes describe the instance.
+	Attributes []ResourceAttribute
+}
+
+// ResourceAttribute describes one particular aspect of a resource,
+// using one (and only one!) of several different semantic models.
+type ResourceAttribute struct {
+    Name string
+    ResourceAttributeModel // inline
+}
+
+// ResourceAttributeModel must have one and only one field set.
+type ResourceAttributeModel struct {
+	Counter *CounterAttribute
+    Metadata *MetadataAttribute
+}
+```
+
+### ResourceClaimParameters
+
+```
+type ResourceClaimParameters struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	metav1.ObjectMeta
+
+	// If this object was created from some other resource, then this links
+	// back to that resource. This field is used to find the in-tree representation
+	// of the claim parameters when the parameter reference of the claim refers
+	// to some unknown type.
+	GeneratedFrom *ResourceClaimParametersObjectReference
+
+    // Shareable indicates whether the allocated claim is meant to be shareable
+    // by multiple consumers at the same time.
+    Shareable bool
+
+	// Parameters contains parameters for semantic models. All of
+	// the parameters specified here must be satisfied before a claim
+	// using these parameters can be allocated.
+    //
+    // At the moment, all of them must be satisfied by the same resource instance.
+	Parameters []ResourceParameter
+}
```

Comment:
Besides the mixing of different "parameters" there's another conceptual aspect that we should consider: in the original design we added the ability to allocate resources from different drivers for a single claim. Each `ResultHandle` then got passed to the right kubelet plugin for the resource that it represents.

I have not replicated that idea in this KEP. I think we should allow it. It's not hard, we just need to add the driver name(s) into the ResourceClaimParameters instead of assuming that it's the same as in the resource class. The resoure class name can still be the default.

Let's decide about one vs. many models per kubelet driver first, then I can update accordingly here.

Reply:
`Requests []ResourceRequest` is now per-driver, so this has been resolved.